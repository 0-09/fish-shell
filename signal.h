/** \file signal.h

The library for various signal related issues

*/

/**
   Get the integer signal value representing the specified signal, or
   -1 of no signal was found
*/
int wcs2sig( const wchar_t *str );

/**
   Get string representation of a signal
*/
const wchar_t *sig2wcs( int sig );

/**
   Returns a description of the specified signal.
*/
const wchar_t *sig_description( int sig );

/**
   Set all signal handlers to SIG_DFL
*/
void signal_reset_handlers();

/**
   Set signal handlers to fish default handlers
*/
void signal_set_handlers();

/**
   Tell fish to catch the specified signal and fire an event, instead of performing the default action
*/
void signal_handle( int sig, int do_handle );
