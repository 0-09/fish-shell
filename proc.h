/** \file proc.h 

    Prototypes for utilities for keeping track of jobs, processes and subshells, as
	well as signal handling functions for tracking children. These
	functions do not themselves launch new processes, the exec library
	will call proc to create representations of the running jobs as
	needed.
	
*/

#ifndef FISH_PROC_H
#define FISH_PROC_H

#include <wchar.h>
#include <signal.h>
#include <unistd.h>

#include "util.h"
#include "io.h"


/**
   Types of internal processes
*/
enum
{
	EXTERNAL,
	INTERNAL_BUILTIN,
	INTERNAL_FUNCTION,
	INTERNAL_BLOCK,
	INTERNAL_EXEC
}
	;


/** 
	A structure representing a single process. Contains variables for
	tracking process state and the process argument list. 
*/
typedef struct process{
	/** argv parameter for for execv */
	wchar_t **argv;
	/** actual command to pass to exec */
	wchar_t *actual_cmd;       
	/** process ID */
	pid_t pid;
	/** 
		Type of process. Can be one of \c EXTERNAL, \c
		INTERNAL_BUILTIN, \c INTERNAL_FUNCTION, \c INTERNAL_BLOCK
	*/
	int type;
	/** File descriptor that pipe output should bind to */
	int pipe_fd;
	/** true if process has completed */
	volatile int completed;
	/** true if process has stopped */
	volatile int stopped;
	/** reported status value */
	volatile int status;
	/** next process in pipeline */
	struct process *next;       	
#ifdef HAVE__PROC_SELF_STAT
	/** Last time of cpu time check */
	struct timeval last_time;
	/** Number of jiffies spent in process at last cpu time check */
	unsigned long last_jiffies;	
#endif
} process_t;


/** Represents a pipeline of one or more processes.  */
typedef struct job
{
	/** command line, used for messages */
	wchar_t *command;              
	/** list of processes in this job */
	process_t *first_process;  
	/** process group ID */
	pid_t pgid;   
	/** true if user was told about stopped job */
	int notified;     
	/** saved terminal modes */
	struct termios tmodes;    
	/** The job id of the job*/
	int job_id;
	/** Whether this job is in the foreground */
	int fg;
	/** 
		Whether the specified job is completely constructed,
		i.e. completely parsed, and every process in the job has been
		forked
	*/
	int constructed;
	/**
	   Whether the specified job is a part of a subshell or some other form of special job that should not be reported
	*/
	int skip_notification;
	
	/** List of IO redrections for the job */
	io_data_t *io;
	
	/** Should the exit status be negated */
	int negate;	
	/** Is this a conditional short circut thing? If so, is it an COND_OR or a COND_AND */
	struct job *next;           
} job_t;

/** Whether we are running a subshell command */
extern int is_subshell;
/** Whether we are running a block of commands */
extern int is_block;
/** Whether we are reading from the keyboard right now*/
extern int is_interactive;
/** Whether this shell is attached to the keyboard at all*/
extern int is_interactive_session;
/** Whether we are a login shell*/
extern int is_login;
/** Whether we are a event handler*/
extern int is_event;
/** Linked list of all jobs */
extern job_t *first_job;   

/**
   Whether a universal variable barrier roundtrip has already been
   made for this command. Such a roundtrip only needs to be done once
   on a given command, unless a unversal variable value is
   changed. Once this has been done, this variable is set to 1, so
   that no more roundtrips need to be done.

   Both setting it to one when it should be zero and the opposite may
   cause concurrency bugs.
*/
extern int proc_had_barrier;

extern pid_t proc_last_bg_pid;

/**
   Sets the status of the last process to exit
*/
void proc_set_last_status( int s );
/**
   Returns the status of the last process to exit
*/
int proc_get_last_status();

/**
   Remove the specified job
*/
void job_free( job_t* j );
/**
   Create a new job
*/
job_t *job_create();

/**
  Return the job with the specified job id.
  If id is -1, return the last job used.
*/
job_t *job_get(int id);

/**
  Return the job with the specified pid.
*/
job_t *job_get_from_pid(int pid);

/**
   Tests if the job is stopped 
 */
int job_is_stopped( const job_t *j );

/**
   Tests if the job has completed
 */
int job_is_completed( const job_t *j );

/**
  Reassume a (possibly) stopped job. Put job j in the foreground.  If
  cont is nonzero, restore the saved terminal modes and send the
  process group a SIGCONT signal to wake it up before we block.

  \param j The job
  \param cont Whether the function should wait for the job to complete before returning
*/
void job_continue( job_t *j, int cont );
/**
   Notify user of nog events.  Notify the user about stopped or
   terminated jobs.  Delete terminated jobs from the active job list.
*/
int job_do_notification();
/**
   Signal handler for SIGCHLD.  Mark any processes with relevant
   information.

*/
void job_handle_signal( int signal, siginfo_t *info, void *con );

/**
   Clean up before exiting
*/
void proc_destroy();


#ifdef HAVE__PROC_SELF_STAT
/**
   Use the procfs filesystem to look up how many jiffies of cpu time
   was used by this process. This function is only available on
   systems with the procfs file entry 'stat', i.e. Linux.
*/
unsigned long proc_get_jiffies( process_t *p );

/**
   Update process time usage for all processes by calling the
   proc_get_jiffies function for every process of every job.
*/
void proc_update_jiffies();

#endif

/**
   Perform a set of simple sanity checks on the job list. This
   includes making sure that only one job is in the foreground, that
   every process is in a valid state, etc.
*/
void proc_sanity_check();

#endif
