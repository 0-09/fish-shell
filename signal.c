/** \file signal.c

The library for various signal related issues

*/

#include "config.h"

#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <signal.h>
#include <dirent.h>
#include <unistd.h>

#include "common.h"
#include "signal.h"
#include "event.h"
#include "reader.h"
#include "proc.h"

/**
   Struct describing an entry for the lookup table used to convert
   between signal names and signal ids, etc.
*/
struct lookup_entry
{
	/**
	   Signal id
	*/
	int signal;
	/**
	   Signal name
	*/
	const wchar_t *name;
	/**
	   Signal description
	*/
	const wchar_t *desc;	
};

/**
   Lookup table used to convert between signal names and signal ids,
   etc.
*/
static struct lookup_entry lookup[] =
{
	{
		SIGHUP, 
		L"SIGHUP",
		L"Terminal hung up"
	}
	,
	{
		SIGINT,
		L"SIGINT",
		L"Quit request from job control (^C)"
	}
	,
	{
		SIGQUIT,
		L"SIGQUIT",
		L"Quit request from job control with core dump (^\\)"
	}
	,
	{
		SIGILL,
		L"SIGILL",
		L"Illegal instruction"
	}
	,
	{
		SIGTRAP, 
		L"SIGTRAP",
		L"Trace or breakpoint trap"
	}
	,
	{
		SIGABRT,
		L"SIGABRT",
		L"Abort"
	}
	,
	{
		SIGBUS, 
		L"SIGBUS",
		L"Misaligned address error"
	}
	,
	{
		SIGFPE, 
		L"SIGFPE",
		L"Floating point exception"
	}
	,
	{
		SIGKILL, 
		L"SIGKILL",
		L"Forced quit"
	}
	,
	{
		SIGUSR1,
		L"SIGUSR1",
		L"User defined signal 1"
	}
	,
	{
		SIGUSR2, L"SIGUSR2",
		L"User defined signal 2"
	}
	,
	{
		SIGSEGV, 
		L"SIGSEGV",
		L"Address boundary error"
	}
	,
	{
		SIGPIPE,
		L"SIGPIPE",
		L"Broken pipe"
	}
	,
	{
		SIGALRM, 
		L"SIGALRM",
		L"Timer expired"
	}
	,
	{
		SIGTERM,
		L"SIGTERM",
		L"Polite quit request"
	}
	,
	{
		SIGCHLD,
		L"SIGCHLD",
		L"Child process status changed"
	}
	,
	{
		SIGCONT,
		L"SIGCONT",
		L"Continue previously stopped process"
	}
	,
	{
		SIGSTOP,
		L"SIGSTOP",
		L"Forced stop"
	}
	,
	{
		SIGTSTP,
		L"SIGTSTP",
		L"Stop request from job control (^Z)"
	}
	,
	{
		SIGTTIN, 
		L"SIGTTIN",
		L"Stop from terminal input"
	}
	,
	{
		SIGTTOU,
		L"SIGTTOU",
		L"Stop from terminal output"
	}
	,
	{
		SIGURG, 
		L"SIGURG",
		L"Urgent socket condition"
	}
	,
	{
		SIGXCPU,
		L"SIGXCPU",
		L"CPU time limit exceeded"
	}
	,
	{
		SIGXFSZ, 
		L"SIGXFSZ",
		L"File size limit exceeded"
	}
	,
	{
		SIGVTALRM,
		L"SIGVTALRM",
		L"Virtual timer expired"
	}
	,
	{
		SIGPROF,
		L"SIGPROF",
		L"Profiling timer expired"
	}
	,
	{
		SIGWINCH,
		L"SIGWINCH",
		L"Window size change"
	}
	,
	{
		SIGIO,
		L"SIGIO",
		L"Asynchronous IO event"
	}
	,
#ifdef SIGPWR
	{
		SIGPWR, 
		L"SIGPWR",
		L"Power failure"
	}
	,
#endif
	{
		SIGSYS, 
		L"SIGSYS",
		L"Bad system call"
	}
	,
	{
		0,
		0,
		0
	}
}
	;


/**
   Test if \c name is a string describing the signal named \c canonical. 
*/
static int match_signal_name( const wchar_t *canonical, 
							  const wchar_t *name )
{
	if( wcsncasecmp( name, L"sig", 3 )==0)
		name +=3;

	return wcscasecmp( canonical+3,name ) == 0;	
}


int wcs2sig( const wchar_t *str )
{
	int i;
	for( i=0; lookup[i].desc ; i++ )
	{
		if( match_signal_name( lookup[i].name, str) )
		{
			return lookup[i].signal;
		}
	}
	return -1;	
}


const wchar_t *sig2wcs( int sig )
{
	int i;
	for( i=0; lookup[i].desc ; i++ )
	{
		if( lookup[i].signal == sig )
		{
			return lookup[i].name;
		}
	}
	return L"Unknown";
}

const wchar_t *sig_description( int sig )
{
	int i;
	for( i=0; lookup[i].desc ; i++ )
	{
		if( lookup[i].signal == sig )
		{
			return lookup[i].desc;
		}
	}
	return L"Unknown";
}

static void default_handler(int signal, siginfo_t *info, void *context)
{
	event_t e;
	e.type=EVENT_SIGNAL;
	e.signal = signal;
	e.function_name=0;

	event_fire( &e, 0 );
}

/**
   Respond to a winch signal by checking the terminal size
*/
static void handle_winch( int sig, siginfo_t *info, void *context )
{
	reader_handle_winch( sig );	
	default_handler( sig, 0, 0 );	
}

/**
   Interactive mode ^C handler. Respond to int signal by setting
   interrupted-flag and stopping all loops and conditionals.
*/
static void handle_int( int sig, siginfo_t *info, void *context )
{
	reader_handle_int( sig );

	default_handler( sig, info, context);	
}

/**
   Sets appropriate signal handlers.
*/
void signal_set_handlers()
{
	struct sigaction act;
	sigemptyset( & act.sa_mask );
	act.sa_flags=SA_SIGINFO;
	act.sa_sigaction = &default_handler;

	/*
	  First reset everything to a use default_handler, a function
	  whose sole action is to fire of an event
	*/
	sigaction( SIGINT, &act, 0);
	sigaction( SIGQUIT, &act, 0);
	sigaction( SIGTSTP, &act, 0);
	sigaction( SIGTTIN, &act, 0);
	sigaction( SIGTTOU, &act, 0);
	sigaction( SIGCHLD, &act, 0);
	sigaction( SIGALRM, &act, 0);

	/*
	  Ignore sigpipe, it is generated if fishd dies, but we can
	  recover.
	*/
	sigaction( SIGPIPE, &act, 0);
	
	if( is_interactive )
	{

		/*
		   Interactive mode. Ignore interactive signals.  We are a
		   shell, we know whats best for the user. ;-)
		*/

		act.sa_handler=SIG_IGN;

		sigaction( SIGINT, &act, 0);
		sigaction( SIGQUIT, &act, 0);
		sigaction( SIGTSTP, &act, 0);
		sigaction( SIGTTIN, &act, 0);
		sigaction( SIGTTOU, &act, 0);

		act.sa_handler = &handle_int;
		act.sa_flags = SA_SIGINFO;
		if( sigaction( SIGINT, &act, 0) )
		{
			wperror( L"sigaction" );
			exit(1);
		}

		act.sa_sigaction = &job_handle_signal;
		act.sa_flags = SA_SIGINFO;
		if( sigaction( SIGCHLD, &act, 0) )
		{
			wperror( L"sigaction" );
			exit(1);
		}

		act.sa_flags = SA_SIGINFO;
		act.sa_handler= &handle_winch;
		if( sigaction( SIGWINCH, &act, 0 ) )
		{
			wperror( L"sigaction" );
			exit(1);
		}

	}
	else
	{
		/*
		  Non-interactive. Ignore interrupt, check exit status of
		  processes to determine result instead.
		*/
		act.sa_handler=SIG_IGN;

		sigaction( SIGINT, &act, 0);
		sigaction( SIGQUIT, &act, 0);

		act.sa_handler=SIG_DFL;

		act.sa_sigaction = &job_handle_signal;
		act.sa_flags = SA_SIGINFO;
		if( sigaction( SIGCHLD, &act, 0) )
		{
			wperror( L"sigaction" );
			exit(1);
		}
	}
}

