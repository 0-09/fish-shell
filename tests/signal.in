function alarm --on-signal ALRM
  echo ALRM received
end


kill -s ALRM $fish_pid

function anychild --on-process-exit 0
  # Type and exit status
  echo $argv[1] $argv[3]
end

command false
command true
command false | command true
