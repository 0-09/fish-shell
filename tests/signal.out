ALRM received
PROCESS_EXIT 1
JOB_EXIT 0
PROCESS_EXIT 0
JOB_EXIT 0
PROCESS_EXIT 1
PROCESS_EXIT 0
JOB_EXIT 0
PROCESS_EXIT 0
